Provide a fast path for filter consists of only "FQN=" and "|"

genlu opened the pull request:
**Scenario:** 
The matching algorithm used for testcase filter is based on a tree structure and is recursive. It is too inefficient (O(n^2)) when used to run a large amount of tests by name. 

For example, running 10k mstest tests using TestCase takes ~4s whereas using filter takes ~150s. As you can see from the screenshot below, in the 10k tests scenario, more than 90% of time is spent in filtering testcases.

![image](https://user-images.githubusercontent.com/788783/30569227-84b9e5f4-9c8e-11e7-992c-97ae24975f77.png)

This change replaces the FilterExpression with a hash table of fully qualified name of test methods, if we detect that only "FullyQualifiedName=" and "|" is used in the filter. As a result, running 10k tests by filter now takes ~15s.

This PR is based on #1052, which should be merged first. And all my change is in bc84847

genlu committed:
Add a fast path for matching filter expression contains only 'FQN=' and '|'

genlu commented:
FYI @AbhitejJohn @peterwald

AbhitejJohn commented:
@genlu : Had a quick chat on this fix internally. Could we make fast filter generic, so other properties could benefit from this as well? The logic we have here does seem quite generic in that we optimize the case where we only have "|" and one property(could be FQN, could be category etc).
/cc @codito , @smadala

AbhitejJohn commented:
An orthogonal query: Is the delta of 11s in execution acceptable? How does this compare to discovery + execution of 10k tests using the traditional approach?

codito commented:
@AbhitejJohn @genlu @peterwald for theory tests can we form a filter expression like `FullyQualifiedName=NS.Class.A | FullyQualifiedName~NS.Class.TheoryB`? In this case, all the child tests for the theory test `NS.Class.TheoryB` should match.

It will be costlier than a regular `=` check, I think filter based execution is anyways a onetime operation. We shouldn't add test framework specific constructs in the platform, it stays forever and impacts other frameworks (learning from TPv1 and mstest days).

peterwald commented:
@AbhitejJohn @genlu @codito I think the `FullyQualifiedName~NS.Class.TheoryB` could work well. When doing execute by name, we'd need to basically run every test this way (theory or not) in order for it to work on all frameworks. 

Basically, it would still be an `=` comparison but in the case of a filter defined with `~`, we'd just need to extract the method name from FullyQualifiedName (remove extra trailing data) before doing the comparison.

I like the idea of using a different comparison operator since it remains backwards compatible with the current functionality.

genlu committed:
Use an internal NFQN filter keyword for execute by name

genlu committed:
initial support for property value regex matching in filter

genlu committed:
Add FilterOptions to TestPlatformOptions

genlu committed:
Add TestPlatformOptions parameter to async TL request sender interface

genlu committed:
Add FilterOptions to TestRunRequestpayload

genlu committed:
Add FilterOptions to RunCriteria

genlu committed:
Pickup filter regex from FilterOptions when using testcase filter

genlu committed:
Fix tests

navin22 commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpression.cs:
```diff
@@ -213,6 +223,13 @@ internal static FilterExpression Parse(string filterString)
 
+            // We try to build a hash table for an filter expression consists of only a single kind of property with 'equal' operation and '|' operator in parallel.
+            // If by the time we finished parsing the entire expression and `isRunByFullyQualifiedName` is still true, then this hash table will be used instead
+            // for "run by fully qualified name".
```
Need to change this comment, to use `canUseFastFilter` ?

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpression.cs:
```diff
@@ -296,19 +334,66 @@ internal static FilterExpression Parse(string filterString)
+                Debug.Assert(this.condition == null && this.left == null && this.right == null);
+
+                if (fastFilterPropertyName.Equals(FilterExpression.NormalizedFullyQualifiedNameFilterKeyword, StringComparison.OrdinalIgnoreCase))
+                {
```
Don't think we would be needing a normalized FQN anymore.

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpressionWrapper.cs:
```diff
@@ -16,26 +18,40 @@ public class FilterExpressionWrapper
+                var regexString = options?.FilterRegEx;
+                if (!string.IsNullOrEmpty(regexString))
+                {
+                    filterRegex = new Regex(regexString);
```
Should we be using `RegexOptions.Compiled`?

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpressionWrapper.cs:
```diff
@@ -16,26 +18,40 @@ public class FilterExpressionWrapper
         /// Initializes FilterExpressionWrapper with given filterString.  
         /// </summary>
-        public FilterExpressionWrapper(string filterString)
+        public FilterExpressionWrapper(string filterString, FilterOptions options = null)
```
Might want to add another constructor since this is a public class :(

AbhitejJohn commented on src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/ITranslationLayerRequestSenderAsync.cs:
```diff
@@ -50,7 +50,7 @@ internal interface ITranslationLayerRequestSenderAsync : IDisposable
         /// Asynchronous equivalent of <see cref="ITranslationLayerRequestSender.StartTestRunWithCustomHost(IEnumerable{string}, string, TestPlatformOptions, ITestRunEventsHandler, ITestHostLauncher)"/>.
         /// </summary>
-        Task StartTestRunWithCustomHostAsync(IEnumerable<string> sources, string runSettings, ITestRunEventsHandler runEventsHandler, ITestHostLauncher customTestHostLauncher);
+        Task StartTestRunWithCustomHostAsync(IEnumerable<string> sources, string runSettings, TestPlatformOptions options, ITestRunEventsHandler runEventsHandler, ITestHostLauncher customTestHostLauncher);
```
Thanks for adding it here too :)

AbhitejJohn commented on src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleWrapper.cs:
```diff
@@ -142,17 +142,12 @@ public void CancelDiscovery()
         public void RunTests(IEnumerable<string> sources, string runSettings, ITestRunEventsHandler testRunEventsHandler)
         {
-            this.RunTests(sources, runSettings, new TestPlatformOptions(), testRunEventsHandler);
+            this.RunTests(sources, runSettings, null, testRunEventsHandler);
```
Do we want users to use the below API when `TestPlatformOptions` is null? Would expect users to use this one in that case.

genlu replied:
We check for null down stream, so I figured there's no harm.

jimmy9988 replied:
Consider: Use named parameters for all these method call with null option.

AbhitejJohn commented:
@Genlu: Looks fairly good to me. Were you able to get the numbers with this flow?

navin22 committed:
Refactored the code.

- Moved Payloads to ObjectModel.Client & Necessary chages in other files.
- When RegEx is null returning the actual value.
- Compiling the regex.

genlu committed:
Refactoring

genlu committed:
Remove NFQN hack

genlu committed:
Add tests for fast filter

genlu commented:
@shyamnamboodiripad @drognanar @navin22 @codito @smadala Could you guys please review this change?

jimmy9988 commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpression.cs:
```diff
@@ -296,19 +294,23 @@ internal static FilterExpression Parse(string filterString)
         /// </summary>
-        /// <param name="propertyValueProvider"> The property Value Provider. </param>
+        /// <param name="propertyValueProvider"> The property Value Provider.</param>
+        /// <param name="propertyValueRegexMatchOpt">Applies RegEx pattern on the property value</param>
```
Should this be removed?

genlu replied:
Yup.

jimmy9988 commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpressionWrapper.cs:
```diff
@@ -16,26 +18,68 @@ public class FilterExpressionWrapper
             {
-                this.filterExpression = FilterExpression.Parse(filterString);
+                // We prefer fast filter when it's available.
+                this.filterExpression = FilterExpression.Parse(filterString, out this.fastFilter);
```
Consider: Make both FilterExpression and FastFilter out parameters or use value tuple.

jimmy9988 commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpressionWrapper.cs:
```diff
@@ -16,26 +18,68 @@ public class FilterExpressionWrapper
+        /// <summary>
+        /// Initializes FilterExpressionWrapper with given filterString.  
+        /// </summary>
+        public FilterExpressionWrapper(string filterString)
```
After adding optional parameter `FilterOptions options = null` for FilterExpressionWrapper, this should be duplicated?

genlu replied:
This is for compatibility, since `FilterExpressionWrapper` is public.

jimmy9988 commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpressionWrapper.cs:
```diff
@@ -60,25 +113,28 @@ public string ParseError
-            return this.filterExpression.Evaluate(propertyValueProvider);
-        }
 
+            return (UseFilterExpression && this.filterExpression.Evaluate(propertyValueProvider)) || 
```
Why the order of UseFilterExpression and UseFastFilter is opposite here?

jimmy9988 commented on src/Microsoft.TestPlatform.Common/Microsoft.TestPlatform.Common.csproj:
```diff
@@ -42,5 +42,9 @@
   <PropertyGroup Label="Configuration">
     <RootNamespace>Microsoft.VisualStudio.TestPlatform.Common</RootNamespace>
   </PropertyGroup>
+
```
Nit: Empty line

jimmy9988 commented on src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleWrapperAsync.cs:
```diff
@@ -14,6 +14,7 @@ namespace Microsoft.TestPlatform.VsTestConsole.TranslationLayer
     using Microsoft.VisualStudio.TestPlatform.ObjectModel;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client.Interfaces;
+    using System;
```
NIT: unnecessary using

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FastFilter.cs:
```diff
@@ -0,0 +1,161 @@
+            }
+            else
+            {
+                throw new ArgumentException();
```
Please add in text defining the support of fast filter in here.

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FastFilter.cs:
```diff
@@ -0,0 +1,161 @@
+                return false;
+            }
+
+            if (PropertyValueRegex != null)
```
nit: Since `Evaluate` would be called for every test case, this if and the `IsFilteredOutWhenMatched` condition could be avoided if the implementation changes to have an inheritance with a Factory to return back the right FastFilter instance. Might just make it faster. Thoughts?

genlu replied:
well, I don't think the logic and overhead here warrants an abstraction like that yet.

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FastFilter.cs:
```diff
@@ -0,0 +1,161 @@
+            return IsFilteredOutWhenMatched ? !matched : matched;
+        }
+
+        private static bool TryGetSinglePropertyValue(string name, Func<string, Object> propertyValueProvider, out string singleValue)
```
nit: Looks like there is only one caller. Can we inline this?

genlu replied:
Sure.

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FastFilter.cs:
```diff
@@ -0,0 +1,161 @@
+            internal void AddOperator(Operator @operator)
+            {
+                if (!containsValidFilter || !(@operator == Operator.And || @operator == Operator.Or))
+                {
```
should `containsValidFilter` be set to false if `!(@operator == Operator.And || @operator == Operator.Or)) ` is satisfied?

genlu replied:
You are right.

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FastFilter.cs:
```diff
@@ -0,0 +1,161 @@
+                    filterHashSetBuilder.Add(condition.Value);
+
+                    // Don't support `Contains`.
+                    if (fastFilterOperation != Operation.Equal && fastFilterOperation != Operation.NotEqual)
```
I guess its because we want to keep the logic tight in case a new operation is added in later.

genlu replied:
Yup.

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpression.cs:
```diff
@@ -50,7 +50,7 @@ internal class FilterExpression
         #region Constructors
 
-        internal FilterExpression()
+        private FilterExpression()
```
I think we can get rid of this.

genlu replied:
Replaced with `FilterExpression(FilterExpression left, FilterExpression right, bool areJoinedByAnd)`

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpression.cs:
```diff
@@ -296,19 +294,23 @@ internal static FilterExpression Parse(string filterString)
                 throw new FormatException(string.Format(CultureInfo.CurrentCulture, CommonResources.TestCaseFilterFormatException, CommonResources.MissingOperator));
             }
 
+            if (fastFilterBuilder.ContainsValidFilter)
```
ToFastFilter already seems to have this check.

genlu replied:
You are right.

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpressionWrapper.cs:
```diff
@@ -16,26 +18,68 @@ public class FilterExpressionWrapper
+        /// Initializes FilterExpressionWrapper with given filterString and options.  
         /// </summary>
-        public FilterExpressionWrapper(string filterString)
+        public FilterExpressionWrapper(string filterString, FilterOptions options = null)
```
We can make this a required parameter now?

genlu replied:
Fixed.

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpressionWrapper.cs:
```diff
@@ -16,26 +18,68 @@ public class FilterExpressionWrapper
+            { 
+                // TODO: surface an error message to user.
+                // We don't match any tests in case of a regex parsing error.
+                this.fastFilter = null;
```
this.ParseError does not surface the message?

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpressionWrapper.cs:
```diff
@@ -60,25 +113,28 @@ public string ParseError
+            {
+                invalidProperties = this.fastFilter.ValidForProperties(supportedProperties);
+            }
+            else if (UseFilterExpression)
```
Do we need a `UseFilterExpression`? Slightly confusing since it seems it is just a mutual exclusion to `UseFastFilter` although not all code paths ensure a null check to `filterexpression`. Maybe we can just make it a null check as before instead?

genlu replied:
I didn't set `ParseError` when there's regex parsing error, in which case, both `filterExpression` and `fastFilter` would be `null`. Seems `TestCaseFilterExpression.validForMatch` would short-circuit validation and evaluation if `ParseError` is set. Will fix.

AbhitejJohn commented on src/Microsoft.TestPlatform.Common/Interfaces/Engine/ClientProtocol/TestExecutionContext.cs:
```diff
@@ -44,7 +45,8 @@ public TestExecutionContext()
             bool isDebug,
-            string testCaseFilter)
+            string testCaseFilter,
+            FilterOptions filterOptions)
```
I'm hoping this would not break older versions of testhost. I do not think so. @navin22 ?

navin22 replied:
I don't think so. 
@genlu Is these scenarios tested ? And also the Object Model changes in VS IDE TestExplorer scenarios need to be verified.

AbhitejJohn commented on test/Microsoft.TestPlatform.Common.UnitTests/Filtering/FastFilterTests.cs:
```diff
@@ -0,0 +1,198 @@
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
+
+    [TestClass]
+    public class FastFilterTests
```
Can we have a test for something like this too please: `FullyQualifiedName=a|(FullyQualifiedName=b|FullyQualifiedName=c)`?

AbhitejJohn commented on test/Microsoft.TestPlatform.Common.UnitTests/Filtering/FastFilterTests.cs:
```diff
@@ -0,0 +1,198 @@
+        }
+
+        [TestMethod]
+        public void FastFilterWithMultipleNotEqualsClauseAndRegex()
```
nit: Please include the expected test outcome in the name.

AbhitejJohn commented:
@dotnet-bot : test Windows_NT / Debug Build please.
@dotnet-bot : test Windows_NT / Release Build please.

AbhitejJohn commented:
@genlu : Just wanted to bring this up here again(We had an offline discussion on this earlier): This would end up executing tests with the same FQN across multiple assemblies. This is a limitation with the filtering logic unfortunately. I do not think this is a prevalent scenario however. 
You might want to validate how this works with MTFM and the new project system though. I believe the source we send over there would only be of one framework - so that should *not* ideally cause problems.
/cc @ManishJayaswal

navin22 commented on src/Microsoft.TestPlatform.Common/Filtering/FilterExpressionWrapper.cs:
```diff
@@ -16,26 +18,68 @@ public class FilterExpressionWrapper
+                    // Property value regex is only supported for fast filter, 
+                    // so we ignore it if no fast filter is constructed.
+
+                    // TODO: surface an error message to suer.
```
What's the error message to surface it here ?

navin22 commented:
@genlu Changes LGTM. Can we add one E2E Acceptance tests in `Microsoft.TestPlatform.AcceptanceTests`

genlu committed:
Add regex replacement string to filter options

genlu committed:
Merge remote-tracking branch 'Microsoft/master' into fastpath

genlu committed:
Address review comments

genlu committed:
Fix fast filter for property with multiple values

genlu commented:
@navin22 I think existing acceptance tests already cover this change. They helped me caught a bug for matching property with multiple values.

navin22 reviewed (approved):
:shipit:

AbhitejJohn committed:
Fixed a PR build issue on unix and potential null ref issue.

AbhitejJohn committed:
Merge branch 'master' of https://github.com/Microsoft/vstest into fastfilterfixes

AbhitejJohn committed:
Merge pull request #3 from AbhitejJohn/fastfilterfixes

Fast filter fixes

AbhitejJohn committed:
A smoke test was failing where Priority which was int could not be converted to a string and singleValue turned out to be null. Mimicked filterExpression.Evaluate.

AbhitejJohn committed:
Merge pull request #4 from AbhitejJohn/morefastpathfixes

More fast path fixes

AbhitejJohn commented:
@dotnet-bot : test Windows_NT / Debug Build please.

AbhitejJohn committed:
Merge branch 'master' into fastpath
